python/ray/data/_internal/execution/bulk_executor.py

```diff
@@ -0,0 +1,109 @@
+import logging
+from typing import Dict, List, Iterator, Optional
+
+import ray
+from ray.data._internal.execution.interfaces import (
+    Executor,
+    ExecutionOptions,
+    RefBundle,
+    PhysicalOperator,
+)
+from ray.data._internal.execution.operators.input_data_buffer import InputDataBuffer
+from ray.data._internal.progress_bar import ProgressBar
+from ray.data._internal.stats import DatasetStats
+
+logger = logging.getLogger(__name__)
+
+
+class BulkExecutor(Executor):
+    """A bulk (BSP) operator executor.
+
+    This implementation emulates the behavior of the legacy Datasets backend. It
+    is intended to be replaced by default by StreamingExecutor in the future.
+    """
+
+    def __init__(self, options: ExecutionOptions):
+        super().__init__(options)
+        self._stats: Optional[DatasetStats] = DatasetStats(stages={}, parent=None)
+        self._executed = False
+
+    def execute(
+        self, dag: PhysicalOperator, initial_stats: Optional[DatasetStats] = None
+    ) -> Iterator[RefBundle]:
+        """Synchronously executes the DAG via bottom-up recursive traversal."""
+
+        assert not self._executed, "Can only call execute once."
+        self._executed = True
+        if not isinstance(dag, InputDataBuffer):
+            logger.info("Executing DAG %s", dag)
+
+        if initial_stats:
+            self._stats = initial_stats
+
+        saved_outputs: Dict[PhysicalOperator, List[RefBundle]] = {}
+
+        def execute_recursive(op: PhysicalOperator) -> List[RefBundle]:
+            # Avoid duplicate executions.
+            if op in saved_outputs:
+                return saved_outputs[op]
+
+            # Compute dependencies.
+            inputs = [execute_recursive(dep) for dep in op.input_dependencies]
+
+            # Fully execute this operator.
+            logger.debug("Executing op %s", op.name)
+            builder = self._stats.child_builder(op.name)
+            try:
+                for i, ref_bundles in enumerate(inputs):
+                    for r in ref_bundles:
+                        op.add_input(r, input_index=i)
+                    op.inputs_done(i)
+                output = _naive_run_until_complete(op)
+            finally:
+                op.shutdown()
+
+            # Cache and return output.
+            saved_outputs[op] = output
+            op_stats = op.get_stats()
+            op_metrics = op.get_metrics()
+            if op_stats:
+                self._stats = builder.build_multistage(op_stats)
+                self._stats.extra_metrics = op_metrics
```

Comment:
Instead of mutating the `extra_metrics` field, adding a `DatasetStats.set_metrics()` setter would make it more clear to readers of the `DatasetStats` class that it is settable.

Reply:
Done.